File: 3975-filter-occupied-intervals/3975-filter-occupied-intervals.py
class Solution(object):
    def filterOccupiedIntervals(self, occupiedIntervals, freeStart, freeEnd):
        """
        :type occupiedIntervals: List[List[int]]
        :type freeStart: int
        :type freeEnd: int
        :rtype: List[List[int]]
        """
        if not occupiedIntervals:
            return []
        occupiedIntervals.sort()
        merged = []
        for start, end in occupiedIntervals:
            if not merged or start > merged[-1][1] + 1:
                merged.append([start, end])
            else:
                merged[-1][1] = max(merged[-1][1], end)
        ans = []
        for start, end in merged:
            if end < freeStart or start > freeEnd:
                ans.append([start, end])
                continue
            if start < freeStart:
                ans.append([start, freeStart - 1])
            if end > freeEnd:
                ans.append([freeEnd + 1, end])

        return ans

File: 3975-filter-occupied-intervals/3975-filter-occupied-intervals.java
class Solution {
    public List<List<Integer>> filterOccupiedIntervals(int[][] occupiedIntervals, int freeStart, int freeEnd) {
        Arrays.sort(occupiedIntervals, (a, b) -> Integer.compare(a[0], b[0]));

        List<int[]> merged = new ArrayList<>();
        merged.add(occupiedIntervals[0]);

        for (int i = 1; i < occupiedIntervals.length; i++) {
            int[] current = occupiedIntervals[i];
            int[] last = merged.get(merged.size() - 1);
            if (last[1] + 1 >= current[0]) {
                last[1] = Math.max(last[1], current[1]);
            } else {
                merged.add(current);
            }
        }
        List<List<Integer>> ans = new ArrayList<>();

        for (int[] interval : merged) {

            int start = interval[0];
            int end = interval[1];

            if (end < freeStart || start > freeEnd) {
                ans.add(Arrays.asList(start, end));
            } 
            else {
                if (start < freeStart) {
                    ans.add(Arrays.asList(start, freeStart - 1));
                }
                if (end > freeEnd) {
                    ans.add(Arrays.asList(freeEnd + 1, end));
                }
            }
        }
        return ans;
    }
}